ngx_http_auth_spnego_module.c

```diff
@@ -952,12 +987,32 @@ ngx_http_auth_spnego_handler(
          * back to basic here... */
         if (NGX_DECLINED == ret) {
             spnego_debug0("GSSAPI failed");
-            return (ctx->ret = NGX_HTTP_FORBIDDEN);
+            if(alcf->allow_basic) {
+                if (NGX_ERROR == ngx_http_auth_spnego_headers_basic_only(r, ctx, alcf)) {
+                    spnego_debug0("Error setting headers");
+                    ctx->ret = NGX_HTTP_INTERNAL_SERVER_ERROR;
+                }
+                else
+                    ctx->ret = NGX_HTTP_UNAUTHORIZED;
+            } else {
+                ctx->ret = NGX_HTTP_FORBIDDEN;
+            }
+            return ctx->ret;
         }
 
         if (!ngx_spnego_authorized_principal(r, &r->headers_in.user, alcf)) {
             spnego_debug0("User not authorized");
-            return (ctx->ret = NGX_HTTP_FORBIDDEN);
+            if(alcf->allow_basic) {
```

Comment:
I don't think adding this here is appropriate.  If the user was successfully authenticated using the provided credentials, but isn't in the list of authorized users, a 403 is correct.  You should not fall back to basic.